Add missing trace log for websocket protocols

laggardkernel opened the pull request:
Formerly we only have trace log for http protocols on `connection_made` and 
`connection_lost` events. **Not very sure**, but I guess we forget
to add trace log for websocket protocols?

laggardkernel committed:
Add missing trace log for websocket protocols

Formerly we only have trace log for http protocol. It seems we missed
to add trace log for websocket protocols.

euri10 reviewed (changes requested):
looks good, this would look better with some tests added, I think we may have some in test_logging.py !

~~now one thing I just realize reading what seem to be a copy of what we do in the http protocols is the following (bear with me it's been a while I did not touch this) : self.logger here is a `uvicorn.error` logger, however iirc when we set `log_level` to trace in Config, we end up with a middleware that spits out those trace log levels on the `uvicorn.asgi` logger. so should we put those under the `uvicorn.asgi` logger ? not sure if that makes sense.~~

no, correcting myself, those are good under that uvicorn.error logger, those are not strictly speaking asgi message where we put the one sent by the middlware

laggardkernel committed:
Add trace logging tests for protocols

Protocols emit trace logs on `connection_made()` and `connection_lost`.
Add missing tests for these logs.

euri10 commented on tests/middleware/test_logging.py:
```diff
@@ -45,6 +46,54 @@ async def test_trace_logging(caplog):
+@pytest.mark.parametrize("protocol", [("h11"), ("httptools")])
+async def test_trace_logging_on_http_protocol(protocol, caplog):
+    config = Config(app=app, log_level="trace", http=protocol)
+    with caplog_for_logger(caplog, "uvicorn.protocol"):
```
```suggestion
    with caplog_for_logger(caplog, "uvicorn.error"):
```
I merged your PR about that, should be good with the "correct" logger name ?

laggardkernel replied:
Yes. Since GH-1086 has been merged, I could set the correct logger name.

euri10 commented on tests/middleware/test_logging.py:
```diff
@@ -45,6 +46,54 @@ async def test_trace_logging(caplog):
 
 
+@pytest.mark.asyncio
+@pytest.mark.parametrize("protocol", [("h11"), ("httptools")])
```
```suggestion
@pytest.mark.parametrize("http_protocol", [("h11"), ("httptools")])
```
small naming variable fight :)

euri10 commented on tests/middleware/test_logging.py:
```diff
@@ -45,6 +46,54 @@ async def test_trace_logging(caplog):
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("protocol", [("websockets"), ("wsproto")])
```
```suggestion
@pytest.mark.parametrize("ws_protocol", [("websockets"), ("wsproto")])
```

euri10 commented on tests/middleware/test_logging.py:
```diff
@@ -45,6 +46,54 @@ async def test_trace_logging(caplog):
+            return websocket.open
+
+    config = Config(app=websocket_app, log_level="trace", ws=protocol)
+    with caplog_for_logger(caplog, "uvicorn.protocol"):
```
same as above for the logger's name

euri10 commented on tests/middleware/test_logging.py:
```diff
@@ -45,6 +46,54 @@ async def test_trace_logging(caplog):
+            if record.name == "uvicorn.error"
+        ]
+        assert any(" - Connection made" in message for message in messages)
+        assert any(" - Connection lost" in message for message in messages)
```
funny, in http test we have one connection made / one connection lost and in ws one we have 2 connection made message, probably the upgrade, not sure.
Not a blocker but I'd like to understand :)

laggardkernel replied:
Yes. There's an upgrade. Http protocol but not ws protocol is set as the default protocol for transport in `handle_http()`.

Maybe we should specify the connection type in protocol trace log.

```python
self.logger.log(TRACE_LOG_LEVEL, "%sHttp connection made", prefix)
# or
self.logger.log(TRACE_LOG_LEVEL, "%sWebsocket connection made", prefix)
```

laggardkernel replied:
Another way to discriminate them is to trace log the "upgrade" request.

euri10 replied:
> Another way to discriminate them is to trace log the "upgrade" request.

would be cool yes, trace should be precise if needed, this is a good disambiguation

euri10 commented on uvicorn/protocols/websockets/websockets_impl.py:
```diff
@@ -7,6 +7,7 @@
 import websockets
 from websockets.extensions.permessage_deflate import ServerPerMessageDeflateFactory
 
+from uvicorn.config import TRACE_LOG_LEVEL
```
```suggestion
from uvicorn.logging import TRACE_LOG_LEVEL
```

euri10 replied:
if someone ca review the relevant PR about that and we merge it first :)

euri10 commented:
genuine question @laggardkernel : why did you need that ?

laggardkernel commented:
> genuine question @laggardkernel : why did you need that ?

What you mean by "that", this PR?

euri10 commented:
> > genuine question @laggardkernel : why did you need that ?
> 
> What you mean by "that", this PR?

yes I mean the ws trace messages :)

laggardkernel commented:
@euri10 Not sure. Because there's only trace log for http protocol. It made me wonder if we forgot to add the same level log for ws.

I did a searching about the history of trace log on http protocol.

Logging on connection made, connection lost was first introduced in uvicorn 0.2.0,
commit 07c6c18. At that time, it was categorized as access log. (In fact, there was
no separate "uvicorn.access" loggers yet.)

Later, with the coming of the `TRACE_LOG_LEVEL` in commit 1238455, http protocol connection logs was changed
to be logged with level 5.

Maybe @tomchristie could answer this, considering he's the guy designed the trace log and later integrated it into `MessageLoggerMiddleware`.

euri10 commented:
Ok I just wondered if there was a use case behind that and if you needed it to debug some sort of issue !

laggardkernel committed:
Specify connection type in protocol trace log

The default transport protocol is http protocol. A ws request
is firstly handled using http protocol, then handled using ws protocol
by switching the protocol. To discriminate the trace log emitter,
we need to specify the protocol type in the log message.

euri10 reviewed (approved):
this looks beautiful, thanks a lot @laggardkernel  and the benefit of having that trace level declared everywhere is that thre's no need to wait for the refactor PR merge !
